src/pytorch/neighbors/getNeighborPairs.py

```diff
@@ -88,39 +104,39 @@ def getNeighborPairs(positions: Tensor, cutoff: float, max_num_neighbors: int =
      tensor([[1., 0., 0.],
              [2., 0., 0.],
              [1., 0., 0.]]),
-     tensor([1., 2., 1.]))
+     tensor([1., 2., 1.]), tensor([3], dtype=torch.int32))
 
     >>> getNeighborPairs(positions, cutoff=1.5) # doctest: +NORMALIZE_WHITESPACE
     (tensor([[ 1, -1,  2],
              [ 0, -1,  1]], dtype=torch.int32),
      tensor([[1., 0., 0.],
              [nan, nan, nan],
              [1., 0., 0.]]),
-     tensor([1., nan, 1.]))
+     tensor([1., nan, 1.]), tensor([3], dtype=torch.int32))
 
     >>> getNeighborPairs(positions, cutoff=3.0, max_num_neighbors=2) # doctest: +NORMALIZE_WHITESPACE
     (tensor([[ 1,  2,  2, -1, -1, -1],
```

Comment:
I don't understand the output.  If `max_num_neighbors` is 2, shouldn't the first return value have shape (2,2) instead of (2,6)?

There seems to be inconsistency in how we interpret `max_num_neighbors`.  In some places it's treated as the total number of neighbors in the list, but in others it's treated as the maximum number of neighbors for a single atom.

Reply:
You are right. max_num_neighbors is **per particle** as per the docs.  But only num_found_pairs<max_num_neighbors*num_atoms is enforced. 
For instance, if 4 particles are at the same point, there are 6 pairs, each particle having 3 neighbors.
However, the function will happily take max_num_neighbors=2, since that results in 8 maximum pairs.

The current build strategy does not provide an easy way to enforce max neighbors per particle.
OTOH changing the interface to receive maximum total pairs instead of per particle would make sense now, but could make it inconvenient if a different strategy is implemented in the future.
Still, the current output format (a compacted list of pairs) would probably require some kind of [partition](https://thrust.github.io/doc/group__partitioning_ga73ee335b49c6ae3dcb019423e268a231.html) if a more sophisticated algorithm would be in place.

Thus, I believe changing the meaning of the input parameter to "maximum number of total pairs" makes sense now.

I added a test that checks that the max_num_neighbors is actually enforced. Naturally, this test fails in cpu and gpu as of now.

So we have two options:
  1. Change the max_num_neighbors agument from "per particle" to max_num_pairs.
  2. Simply change the documentation to reflect that, while an atom might have more than max_num_neighbors, this is not an error unless num_pairs>max_num_neighbors*num_atoms

What do you think?
cc @raimis

Reply:
Given how the code is written, I think a limit on the total pairs is what we need.  The reason for having this argument at all is that we need to know what size to make the output tensors.  And nothing in the output tensors depends on how many neighbors a particular atom has.  They just depend on the total number of pairs in the neighbor list.

Reply:
I changed the meaning and name from max_num_neighbors (maximum number of neighbors per particle) to max_num_pairs (maximum number of total pairs).
I modified the tests to reflect this.